src/Fdc3.AppDirectory/AppDirectoryExtensions.cs

```
@@ -0,0 +1,39 @@
+﻿/*
+ * Morgan Stanley makes this available to you under the Apache License,
+ * Version 2.0 (the "License"). You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * See the NOTICE file distributed with this work for additional information
+ * regarding copyright ownership. Unless required by applicable law or agreed
+ * to in writing, software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions
+ * and limitations under the License.
+ */
+
+using System.Threading.Tasks;
+
+namespace MorganStanley.Fdc3.AppDirectory
+{
+    public static class AppDirectoryExtensions
+    {
+        /// <summary>
+        /// Returns an application by appId from the AppDirectory, or <c>null</c> if the app is not found.
+        /// </summary>
+        /// <param name="appDirectory">The <see cref="IAppDirectory"/> instance</param>
+        /// <param name="appId">Application identifier</param>
+        /// <returns>The application or <c>null</c></returns>
+        public static async Task<Fdc3App?> TryGetApp(this IAppDirectory appDirectory, string appId)
+        {
+            try
```

I think it would be better to avoid the exception in a `TryGet*` method rather than catching it. Would that be possible?

---

The change is about throwing instead of returning `null` - which we can debate, see the linked issue. I don't feel strongly about it, I just noticed that the AppD REST API returns an error response when the app is not found.

---

Yeah I agree with the linked issue & the `Get*` part, my comment is just for the `TryGet*` method.

https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/exceptions-and-performance#try-parse-pattern
In my mind if I see an API which provides both a `Get*` and a `TryGet*` method my assumption is that the first one can throw an exception but the later won't. If the only way of implementing the `TryGet*` is by wrapping the `Get*` in a try-catch then I don't think it should be part of the API - the consumers can do that themselves and decide better what to do in the catch block.

---

This is a helper method in case you really just want to swallow the exception. `GetApp` shouldn't be called with arbitrary identifiers anyway. The most likely scenario when you would try to fetch the API with a nonexistent appId is when an app is removed from the AppD but some persisted layout tries to load it. I'm fine with removing this extension method altogether.

---

Should we just suggest consumers use ContinueWith()?

---

Do you mean this?
```cs
appDirectory.GetApp(appId).ContinueWith(t => t.IsFaulted ? null : t.Result)
```
That's the same thing just with more boilerplate.

---

I think the confusion is the Try* pattern is around returning a bool and out.  For this to work with TPL you will have a difficult tuple so just embrace using TPL.  Perhaps we split this one item out of the rest of the PR so the rest can merged.

---

Removed the extension.